src/main/java/org/openrewrite/github/PreferTemurinDistributions.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2023 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.github;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.HasSourcePath;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.yaml.YamlIsoVisitor;
+import org.openrewrite.yaml.YamlVisitor;
+import org.openrewrite.yaml.tree.Yaml;
+
+import java.time.Duration;
+
+import static org.openrewrite.github.ActionsSetupJavaAdoptOpenJDKToTemurin.DISTRIBUTION_MATCHER;
+
+public class PreferTemurinDistributions extends Recipe {
+    @Override
+    public String getDisplayName() {
+        return "Use `actions/setup-java` `temurin` distribution as they are cached in hosted runners";
+    }
+
+    @Override
+    public String getDescription() {
+        return "[Host runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources/) include Temurin by default as part of the (hosted-tool-cache)(https://github.com/actions/setup-java/blob/main/docs/advanced-usage.md#hosted-tool-cache)." +
+                "\n. Using Temurin speeds up builds as there is no need to download and configure the Java SDK with every build";
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(1);
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new HasSourcePath<>(".github/workflows/*.yml");
+    }
+
+    @Override
+    protected YamlVisitor<ExecutionContext> getVisitor() {
+        return new UseTemurinVisitor();
+    }
+
+    private static class UseTemurinVisitor extends YamlIsoVisitor<ExecutionContext> {
+        @Override
+        public Yaml.Mapping.Entry visitMappingEntry(Yaml.Mapping.Entry entry, ExecutionContext ctx) {
+            if (DISTRIBUTION_MATCHER.matches(getCursor()) && !((Yaml.Scalar) entry.getValue()).getValue().equals("temurin")) {
+                return super.visitMappingEntry(entry.withValue(((Yaml.Scalar) entry.getValue()).withValue("temurin")), ctx);
+            }
+            return super.visitMappingEntry(entry, ctx);
+        }
+    }
```

Comment:
This recipe only helps if it is applied to hosted runners but it should not hurt if it is applied to private runners 

We can try using this documentation https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources as a filter but it might not be necessary.

Reply:
What do you think @timtebeek

Reply:
Looking at the list of supported runners the labels seem rather predictable. Perhaps we can be extra safe here and use a regex to capture current and future runners, and only apply then? Something like `^(windows|ubuntu|macos)-(latest|\d+(\.\d+)?)$` seems to be quite selective.

That way we are unlikely to change the JDK when users are already using custom runners, as they might then also be likely to use a specific known distribution intentionally.

Reply:
That makes sense. Thank you for the insights

Do you have any example recipe that uses a previous key to make a decision? 

I pushed https://github.com/openrewrite/rewrite-github-actions/pull/50/commits/aef03daf35894cd6a0844b78724af00cdc47fee9to illustrate what I mean, and I am not sure if that pattern would be acceptable

It seems to work based on the tests but it might be not be idiomatic or safe to do it that way with the shared static reference

Reply:
I'm no expert here, but what I would have expected was a custom visitor in `getSingleSourceApplicableTest`, which looks for the presence of any external runner, and if none are found add a marker that indicates it's safe to execute the recipe. Only when all such applicability test recipes output a marker (or change) are the subsequent recipe steps executed, I believe. That's the typical way to do these filter steps from what I've seen, but I don't know if there's a sample for that in Yaml recipes already. Does such an approach make sense to you?

Reply:
Can you point me to any example that does this? (even if it is not for Yaml)